### PR TITLE
[Certora] Reentrancy

### PR DESCRIPTION
**Pull request by QGarchery**
This PR's goals is to prove that there is no possibility of a reentrancy attacks, assuming that:
- Tokens are trusted. This assumption could potentially be relaxed, based on the fact that the accounting does not depend on reentrant calls. In fact, MetaMorpho has been built with this concern in mind, notably to take into account ERC777 tokens. To simplify the verification though, tokens are assumed to not reenter the contract.
- Morpho Blue and Morpho Blue markets are trusted. Morpho Blue is an immutable contract, and thoroughly verified.

Note also that both assumption make sense in the context of timelocked markets: a market (and its corresponding collateral and loan token) can only be added after going through a timelock phase. After a proper analysis during this period, a market could be considered safe

### CODE REVIEW

File: certora/specs/Reentrancy.spec [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+methods {
+    function multicall(bytes[]) external returns(bytes[]) => NONDET DELETE;
+
+    function _.accrueInterest(MetaMorphoHarness.MarketParams) external => voidSummary() expect void;
+    function _.supply(MetaMorphoHarness.MarketParams, uint256, uint256, address, bytes) external => uintPairSummary() expect (uint256, uint256);
+    function _.withdraw(MetaMorphoHarness.MarketParams, uint256, uint256, address, address) external => uintPairSummary() expect (uint256, uint256);
+
+    function _.transfer(address, uint256) external => boolSummary() expect bool;
+    function _.transferFrom(address, address, uint256) external => boolSummary() expect bool;
+    function _.balanceOf(address) external => uintSummary() expect uint256;
```
Comment (MathisGD) [inline certora/specs/Reentrancy.spec:11]: would it be worth having comments here saying what we expect from external contracts (erc20 isn't reentrant, morpho isn't reentrant)?

Comment (MathisGD) [reply certora/specs/Reentrancy.spec:11]: actually it's written later
```
// Ignore calls to tokens and Morpho markets as they are trusted (they have gone through a timelock).
```

File: certora/specs/Reentrancy.spec [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+function boolSummary() returns bool {
+    ignoredCall = true;
+    bool value;
+    return value;
+}
+
+persistent ghost bool ignoredCall;
+persistent ghost bool hasCall;
```
Comment (MathisGD) [inline certora/specs/Reentrancy.spec:38]: I think that we could have only one "unsafeCall", which is turned true on all calls and turned back to false by the summaries of trusted contracts, but I don't know if it's really better

Comment (MathisGD) [reply certora/specs/Reentrancy.spec:38]: ah no actually the CALL hook is "executed" after the summary?

**Review by MerlinEgalite — APPROVED** [review]

File: certora/specs/Reentrancy.spec (hunk 2 above)
Comment (QGarchery) [reply certora/specs/Reentrancy.spec:38]: Yes it is executed after the summary
Screenshot_2024-02-02_11-03-12

File: certora/specs/Reentrancy.spec [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+hook CALL(uint g, address addr, uint value, uint argsOffset, uint argsLength, uint retOffset, uint retLength) uint rc {
+    if (ignoredCall) {
+        // Ignore calls to tokens and Morpho markets as they are trusted (they have gone through a timelock).
+        ignoredCall = false;
+    } else {
+        hasCall = true;
+    }
```
Comment (MathisGD) [inline certora/specs/Reentrancy.spec:46]: useless suggestion ^^
```suggestion
    // Ignore calls to tokens and Morpho markets as they are trusted (they have gone through a timelock).
    hasCall = !ignoredCall;
    ignoredCall = false;
```

**Review by MathisGD — APPROVED** [review]

File: certora/specs/Reentrancy.spec (hunk 3 above)
Comment (QGarchery) [reply certora/specs/Reentrancy.spec:46]: I think you can't do that because otherwise having an ignored call after a non-ignored call will not be caught

Comment (MathisGD) [reply certora/specs/Reentrancy.spec:46]: nice catch
```suggestion
    // Ignore calls to tokens and Morpho markets as they are trusted (they have gone through a timelock).
    if (!ignoredCall) {
        hasCall = true;  
    }
    ignoredCall = false;
```

Comment (QGarchery) [reply certora/specs/Reentrancy.spec:46]: It would work too, merging as is for now

**Pull request merged**